cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java

```diff
@@ -16,18 +16,136 @@
 
 package co.cask.cdap.api.workflow;
 
+import java.util.List;
 import java.util.Map;
 import javax.annotation.Nullable;
 
 /**
- * Interface to represent the data that is transferred from one node to the next node in the {@link Workflow}
+ * Interface to represent the data that is transferred from one node to the next nodes in the {@link Workflow}.
  */
 public interface WorkflowToken {
+
+  /**
+   * Put the specified key-value entry in the {@link WorkflowToken}.
+   * The token may store additional information about the context in which
+   * this key is being set, for example, the unique name of the workflow node.
+   * @param key   the key representing the entry
+   * @param value the value for the key
+   */
+  void put(String key, String value);
+
   /**
-   * Get the Hadoop counters from the previous MapReduce program in the Workflow. The method returns null
-   * if the counters are not set.
+   * Same key can be added to the WorkflowToken by multiple nodes.
+   * This method returns the {@link List} of {@link NodeValueEntry}, where
+   * each entry represents the unique node name and the value that it set
+   * for the specified key.
+   * <p>
+   * The list maintains the order in which the values were
+   * inserted in the WorkflowToken for a specific key except in the case of fork
+   * and join. In case of fork in the Workflow, copies of the WorkflowToken are made
+   * and passed it along the each branch. At the join, all copies of the
```

Comment:
I would not go into that much detail, as it limits us as to how we can change this in the future. All I would say here is that order is guaranteed for nodes within the same branch, but not across different branches.